### PR TITLE
Preprocessor: use hidesets to manage token expansion disabling

### LINKED ISSUE #380: Preprocessor: expansion disabled when it shouldn't be
```c
#define foo(X) 1 bar
#define bar(X) 2 foo

foo(X)(Y)(Z)
```

gcc and clang expand this to:
`1 2 1 bar`

aro expands it to:
`1 2 foo(Z)`

### PR DESCRIPTION
**Pull request by ehaas**
Based on https://www.spinellis.gr/blog/20060626/cpp.algo.pdf

Closes #380

### CODE REVIEW

**Comment by ehaas** [comment]
It's late and I need some sleep but I think this is working; feel free to give it a look but I'm definitely going to want to document it more. It's a bit slower than status quo but seems to follow what clang/gcc do and is able to make more progress on `zig2.c` (now that is running into some invalid type problems in the parser, possibly related to missing definitions somehwere)

```
Benchmark 1 (24 runs): /home/ehaas/source/arocc/zig-out/bin/arocc-master -I. -fsyntax-only sqlite3.c
  measurement          mean ± σ            min … max           outliers         delta
  wall_time           213ms ±  580us     212ms …  214ms          1 ( 4%)        0%
  peak_rss           62.3MB ± 48.8KB    62.2MB … 62.4MB          0 ( 0%)        0%
  cpu_cycles          782M  ± 1.69M      780M  …  785M           0 ( 0%)        0%
  instructions       1.67G  ± 6.23      1.67G  … 1.67G           1 ( 4%)        0%
  cache_references   2.08M  ± 26.9K     2.04M  … 2.11M           0 ( 0%)        0%
  cache_misses        540K  ± 9.90K      515K  …  557K           0 ( 0%)        0%
  branch_misses      7.53M  ± 7.63K     7.52M  … 7.54M           0 ( 0%)        0%
Benchmark 2 (24 runs): /home/ehaas/source/arocc/zig-out/bin/arocc-hideset -I. -fsyntax-only sqlite3.c
  measurement          mean ± σ            min … max           outliers         delta
  wall_time           216ms ±  697us     215ms …  218ms          1 ( 4%)        💩+  1.7% ±  0.2%
  peak_rss           62.5MB ± 49.4KB    62.4MB … 62.5MB          0 ( 0%)          +  0.3% ±  0.0%
  cpu_cycles          797M  ± 2.50M      793M  …  803M           0 ( 0%)        💩+  1.9% ±  0.2%
  instructions       1.71G  ± 7.30      1.71G  … 1.71G           0 ( 0%)        💩+  2.5% ±  0.0%
  cache_references   2.05M  ± 13.3K     2.03M  … 2.10M           1 ( 4%)          -  1.1% ±  0.6%
  cache_misses        560K  ± 7.56K      549K  …  579K           1 ( 4%)        💩+  3.9% ±  1.0%
  branch_misses      7.57M  ± 8.04K     7.55M  … 7.58M           0 ( 0%)          +  0.6% ±  0.1%
```

**Comment by ehaas** [comment]
Updated version benchmark looks similar
```
Benchmark 1 (24 runs): /home/ehaas/source/arocc/zig-out/bin/arocc-master -I. -fsyntax-only sqlite3.c
  measurement          mean ± σ            min … max           outliers         delta
  wall_time           213ms ±  859us     211ms …  216ms          3 (13%)        0%
  peak_rss           62.3MB ± 49.9KB    62.2MB … 62.4MB          0 ( 0%)        0%
  cpu_cycles          782M  ± 1.80M      778M  …  786M           1 ( 4%)        0%
  instructions       1.67G  ± 2.08      1.67G  … 1.67G           0 ( 0%)        0%
  cache_references   2.08M  ± 29.5K     2.04M  … 2.12M           0 ( 0%)        0%
  cache_misses        542K  ± 10.3K      512K  …  556K           1 ( 4%)        0%
  branch_misses      7.53M  ± 8.23K     7.51M  … 7.55M           0 ( 0%)        0%
Benchmark 2 (24 runs): /home/ehaas/source/arocc/zig-out/bin/arocc-hideset -I. -fsyntax-only sqlite3.c
  measurement          mean ± σ            min … max           outliers         delta
  wall_time           215ms ± 1.32ms     214ms …  221ms          1 ( 4%)          +  1.2% ±  0.3%
  peak_rss           62.4MB ± 47.0KB    62.3MB … 62.5MB          0 ( 0%)          +  0.2% ±  0.0%
  cpu_cycles          793M  ± 2.37M      788M  …  798M           0 ( 0%)        💩+  1.3% ±  0.2%
  instructions       1.72G  ± 2.46      1.72G  … 1.72G           0 ( 0%)        💩+  2.9% ±  0.0%
  cache_references   2.09M  ± 28.3K     2.05M  … 2.14M           0 ( 0%)          +  0.6% ±  0.8%
  cache_misses        562K  ± 14.9K      546K  …  608K           4 (17%)        💩+  3.7% ±  1.4%
  branch_misses      7.48M  ± 7.66K     7.47M  … 7.50M           1 ( 4%)          -  0.6% ±  0.1%
```

File: src/aro/Preprocessor.zig [hunk 1]
```diff
@@ -1884,22 +1890,6 @@ fn expandVaOpt(
     }
 }
 
-fn shouldExpand(tok: Token, macro: *Macro) bool {
```
Comment (Vexu) [inline src/aro/Preprocessor.zig:1887]: Removing this makes the `start` and `end` fields of `Macro` unused.

File: src/aro/Hideset.zig [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+    if (a == .sentinel or b == .sentinel) return .sentinel;
+    self.intersection_map.clearRetainingCapacity();
+
+    var cur: Index = .sentinel;
+    var head: Index = .sentinel;
+    var it = self.iterator(a);
+    var a_len: usize = 0;
+    while (it.next()) |identifier| : (a_len += 1) {
+        const str = identifier.slice(self.comp);
+        try self.intersection_map.put(self.comp.gpa, str, {});
```
Comment (Vexu) [inline src/aro/Hideset.zig:168]: Is the string equality necessary here? I tested using `Hideset.Identifier` as the key and all tests still passed.

**Review by Vexu — APPROVED** [review]
Thanks for working on this!

File: src/aro/Hideset.zig (hunk 2 above)
Comment (ehaas) [reply src/aro/Hideset.zig]: I think that's right, updated.

**Comment by ehaas** [comment]
I think this change brings us closer to clang/gcc compliance but the final boss of preprocessor stress tests still remains: https://github.com/Hirrolot/metalang99

I get different results from `tests/eval/rec.c` but I haven't managed to reduce it to a minimal example yet.

**Pull request merged**